src/libespeak-ng/translate.c

```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
-		if (c_lower == (word & 0xffff)) {
-			if ((word >> 16) == 0) {
-				new_c = replace_chars[ix+1];
-				break;
-			}
-			if ((word >> 16) == (unsigned int)towlower2(next_in, tr)) {
-				new_c = replace_chars[ix+1];
-				ignore_next = true;
-				break;
-			}
-		}
+	// if requested, print trace
+	if (option_phonemes & espeakPHONEMES_TRACE) {
+		fprintf(f_trans, "Replace   %s > %s\n", charfrom, charto);
 	}
+}
 
-	if (new_c == 0)
-		return c; // no substitution
+static void TranslateChars(Translator *tr, char *source)
+{
+	// Replace characters using Translator.replace_chars in passed source buffer
+	int *replace_chars = tr->langopts.replace_chars;
+	char *mysource = source;
+	char *startmatch;
+	int *from;
+	int *to;
+	bool upper_case = false;
+
+	if ((from = tr->langopts.replace_chars) == NULL)
+		return;
 
-	if (new_c & 0xffe00000) {
-		// there is a second character to be inserted
-		// don't convert the case of the second character unless the next letter is also upper case
-		c2 = new_c >> 16;
-		if (upper_case && iswupper(next_in))
-			c2 = ucd_toupper(c2);
-		*insert = c2;
-		new_c &= 0xffff;
+	char *curchar = source; // pointer to the source clause
+	int curint;             // current UTF-16 character
+	int used;               // bytes used in buffer
+	int *curfrom;           // pointer to current 'from'
+	bool next_step = false; // signal to jumping to next 'from' part in dictionary
+	while (GetNextTranslateChars(&from, &to, next_step)) {
+		next_step = true;
+		curfrom = from;
+		curchar = source;
+		startmatch = 0;
+		curint = 0;
+		do {
+			used = utf8_in(&curint, curchar);
+			if (curint < ' ') {        // skip control characters
+				curchar += used;
+				continue;
+			}
+			if ((upper_case = iswupper(curint)) == true) { // remember, if uppercase
+				curint = ucd_tolower(curint);
+			};
+			if (curint == *curfrom) {  // if matches
+				if (startmatch == 0)
+					startmatch = curchar;
+				curfrom++;             // move check to next character
+				if (*curfrom == ' ') { // if full match, replace
+					ReplaceMatchingChars(&startmatch, from, to, upper_case);
+					startmatch = 0;    // reset match to start of 'from'
+					curfrom = from;
+				}
+			} else {                   // reset match to start of 'from'
+				startmatch = 0;
+				curfrom = from;
+			}
+			curchar += used;
+		} while (curint > 0);
 	}
-
-	if (upper_case)
-		new_c = ucd_toupper(new_c);
-
-	*wordflags |= FLAG_CHAR_REPLACED;
-	if (option_phonemes & espeakPHONEMES_TRACE) {
-		char msg[21] = {'R','e','p','l','a','c','e',':',' '};
-		char *index = &msg;
-		index +=9;
-        index += utf8_out(c, index);
-		*index++ = ' ';
-		*index++ = '>';
-		*index++ = ' ';
-		index += utf8_out(new_c, index);
-		index += utf8_out(c2, index);
-		*index = 0;
-		fprintf(f_trans, "%s\n", msg);
-	}
-	return new_c;
 }
 
 static int TranslateChar(Translator *tr, char *ptr, int prev_in, unsigned int c, unsigned int next_in, int *insert, int *wordflags)
```

Comment:
`wordflags` is now no longer used (after removing the `SubstituteChar` call). Therefore, it should be removed.
```
src/libespeak-ng/translate.c:1948:122: warning: unused parameter ‘wordflags’ [-Wunused-parameter]
  int TranslateChar(Translator *tr, char *ptr, int prev_in, unsigned int c, unsigned int next_in, int *insert, int *wordflags)
                                                                                                                    ^~~~~~~~~
```